prometheus-net/Histogram.cs

```diff
@@ -48,55 +47,35 @@ internal Histogram(string name, string help, string[] labelNames, double[] bucke
 
         public class Child : Advanced.Child, IHistogram
         {
-            private double _sum = 0;
-            private ulong _count = 0;
-            private ulong[] _bucketCounts;
-            private readonly object _lock = new object();
-
-            //re-using WireFormat.Histogram to avoid excessive array allocations 
-            //TODO: this will break thread-safety if the very same histogram instance is registered on multiple CollectorRegistries
-            //TODO: we should investigate if HdrHistogram can be a good candidate to use here internally - that's heavily optimized
-
-            private Advanced.DataContracts.Histogram _wireMetric;
-            private double[] _buckets;
+            private ThreadSafeDouble _sum = new ThreadSafeDouble(0.0D);
+            private ThreadSafeLong[] _bucketCounts;
```

Comment:
Even though we the data contract is using ulong here long will be sufficient.

The java client actually uses double for this count, but because of the precision on a double in java it will lose the ability to increment by 1 at a lower value than the long will. So technically, our long can "count" higher than the official java client can. That said, we could potentially rollover, so we may want to guard that, but we won't rollover until 9,223,372,036,854,775,808.  If we ran a program continuously for 100 years how many observations per second is that? 9,223,372,036,854,775,808 / (60 * 60 * 24 * 365 * 100) = 2.9 billion observations /second. I think we can safely assume that will not reasonably happen so I think we are safe to not worry about rollover here.